rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,526 @@
+});
+```
+
+### Internal properties
+
+Internal properties (`defaultProps`) are the properties provided by the `useState` hook, that should be overridden by
+External properties, those are the basic properties that ensure proper functioning of a slot but allowing the user
+to opt out of them.
+
+They can be declared when resolving the provided shorthand of a slot (`props.button`, in this example):
+
+```ts
+// useAccordionHeader
+const buttonShorthand = useARIAButtonShorthand(props.button, {
+  required: true,
+  // internal properties
+  defaultProps: {
+    disabled,
+    disabledFocusable,
+    'aria-expanded': open,
+    type: 'button',
+  },
+});
+```
+
+### Overrides
+
+Overrides are the properties provided by the `render` method, that should override all the other properties.
+On most of the implementations of slots we have so far in v9 we tend to avoid using overrides, except for one case that is
+spread all over the components, which is `children` override:
+
+```tsx
+// renderAccordionHeader
+
+<slots.root {...slotProps.root}>
+  <slots.button {...slotProps.button}>
+    {state.expandIconPosition === 'start' && <slots.expandIcon {...slotProps.expandIcon} />}
+    {slots.icon && <slots.icon {...slotProps.icon} />}
+    {slotProps.root.children}
+    {state.expandIconPosition === 'end' && <slots.expandIcon {...slotProps.expandIcon} />}
+  </slots.button>
+</slots.root>
+```
+
+In this example we provide to both `root` and `button` slots overrides of their children value, ignoring
+any previous values that was passed either by Internal properties or External properties. Slots that tend to override
+their children are normally referred to as `wrappers` since they're normally used for the case of
+wrapping other slots inside of them.
+
+## Current issues
+
+[Slot children render function won't override existing children. #27089](https://github.com/microsoft/fluentui/issues/27089)
+
+Repro: https://codesandbox.io/s/elated-babbage-46ez1w?file=/example.tsx
+
+It's not possible to use children render function to override an entire slot, if `children` are already used. The above is a minimal repro of an issue that can be seen in Fluent in the `AccordionHeader` component. The component renders the `button` slot which contains its own children. JSX children will win over props children - which is how children render functions are rendered by `getSlots`
+
+https://github.com/microsoft/fluentui/blob/220b321042b4bcc126c86a91811a34b04934d49b/packages/react-components/react-accordion/src/components/AccordionHeader/renderAccordionHeader.tsx#L13-L25
+
+Requirements for children render function:
+
+- should be able to override the entire slot
+- When rendering the original children, should respect nested slots.
+
+## Problem statement
+
+The main problem revolves around coalescing those 3 instances of properties a slot might have! At the moment,
+our current slot declaration mechanism revolves around 2 separate methods used in 2 different moments:
```

Comment:
> does not work in a case when a render template has explicit children

It doesn't work on the case for implicit children also. take `expandIcon` slot on `AccordionHeader` as an example, we don't have an explicit children on the render template, but one is provided by `defaultProps` and is lost since is competing for the same place as the render function itself.

https://github.com/microsoft/fluentui/blob/71b00883c2c3080f13232cbcea92105cf4d70d54/packages/react-components/react-accordion/src/components/AccordionHeader/useAccordionHeader.tsx#L67-L73